Refactor to-do list with redux

HongJungKim-dev opened the pull request:
초기 커밋입니다.

HongJungKim-dev committed:
Add initial commit

초기 커밋입니다.

moonkii commented on src/index.jsx:
```diff
@@ -1,9 +1,16 @@
+
 ReactDOM.render(
-  <App />,
+  (
+    <Provider store={store}>
+      <App />
+    </Provider>
+  ),
```
`store` 는 당장 필요한가요?

TDD 싸이클을 제대로 하다보면 필요를 찾아가면서 개발을 진행하게 돼요.

항상 테스트 코드를 먼저 작성하는 습관을 하면 좋아요 :)

HongJungKim-dev replied:
강의를 듣고 따라한것을 일단 커밋했었는데 다시 처음부터 시작하겠습니다.

moonkii reviewed (commented):
![ok](https://user-images.githubusercontent.com/46776329/107959145-57bb8380-6fe6-11eb-9c36-3aaf6cc6f6d2.png)

과정을 작게작게 하나씩 밝아 나가는 게 좋은 것 같아요!

TDD 사이클을 잘 지키면서 필요할 때 코드를 바꿔나간다 생각하시면 좋을 것 같아요 :)

HongJungKim-dev committed:
Initial state

과제를 시작하기 위해 처음 과제 받은 상태로 만들었습니다.

HongJungKim-dev commented:
#2-16 회고
어제 아무것도 하지 못했습니다. 지난주에도 무기력이 이어졌었는데 나름 집에서 쉰다고 쉬면서 공부도 하면서 버텼는데 이번주에 더이상 버티기 힘들었습니다. 집중하려 해도 아무것도 집중못하다가 코로나로 인해 집에만 있는것이 가장 큰 원인인것 같다고 판단했습니다. 그래서 어제 오랜만에 친구들과 통화로 연락해봤습니다. 상태는 나아졌지만 햇빛을 맞으며 밖에 나가는 활동이 필요했습니다. 그래서 다음날이 되면 낮에 나가서 햇빛을 맞으면서 산책을 해야겠다고 다짐했습니다.

#2-17 회고
막상 나가려니 춥고 공부할게 많다는 이유로 그냥 바로 공부를 하려고 했습니다. 1~2시간있다가 도저히 안되겠다 싶어서 야외에 나가서 햇빛을 맞으며 산책하고 왔습니다. 훨씬 나아졌고 다시 공부에 집중이 되었습니다.
코로나로 인한 영향이 생각보다 큰것 같습니다. 제때 끼니를 챙겨먹고 햇빛을 맞으며 주기적으로 운동하는등 기본적인 것을 놓치다보니 더 큰 어려움으로 돌아왔습니다. 이제 기본적인것들은 바쁘더라도 꼭 필요한것으로 인식하고 시간을 내서라도 해야겠습니다.

moonkii commented:
코로나 때문에 어려움을 겪고 계시군요 🤔 

왜 공부가 잘 안되고 집중이 잘 안되는지 원인을 찾으려는 고민을 중요한 것 같아요.

혼자 공부를 하고 계신가요? 아무래도 혼자하면 고립된 느낌과 스스로 잘 하고 있는지 어떤지 알기 어려운 경우가 많아요.

가능하면 온라인 스터디라도 진행해보시는건 어떠신가요?

아니면 코드숨 디스코드 방도 있던데 공부할 때 들어가서 함께 공부하는 사람들과 이야기도 나눠보는건 어떨까요?

스스로 능동적으로 움직이지 않으면 변화가 일어나지 않아요.

HongJungKim-dev commented:
온라인 스터디라면 어떤것일까요? 자바스크립트 스터디는 처음에 참여하다가 코드숨 과제로도 벅차서 참여하지 못하게되었습니다...

HongJungKim-dev committed:
Add reducer test code about update input

입력에 따른 업데이트에 대해 reducer 테스트 코드를 추가합니다.

HongJungKim-dev commented:
다음 에러를 구글에 검색해도 나오지 않아서 아직 해결하지 못했습니다.
혹시 initial commit에서 강의를 따라하던 파일들을 지우는 과정에서 문제가 생긴것인지 모르겠습니다.

![image](https://user-images.githubusercontent.com/58525009/108242787-ab69d080-7190-11eb-970d-9b95bfe5ebcd.png)

HongJungKim-dev committed:
Add reducer test code of add and delete case todo

할일을 추가하거나 삭제하는 경우의 reducer test code를 추가했습니다.

HongJungKim-dev committed:
Add reducer test code of update input

입력에 따른 업데이트에 대해 reducer 테스트 코드를 추가합니다.

HongJungKim-dev committed:
Add reducer of update input

입력에 따른 업데이트에 대해 reducer 코드를 구현했습니다.

HongJungKim-dev committed:
Add reducer test code of adding todo

할일을 추가하는 경우의 reducer test code를 추가했습니다.

HongJungKim-dev committed:
Add reducer test code of adding and deleting todo

할일을 추가하거나 삭제하는 경우의 reducer test code를 추가했습니다.

HongJungKim-dev commented:
d897707f52d0f95c55a590da299af6a9eea0e417
이 커밋은 신경안쓰셔도 됩니다.
cddf99fba9a1ce86c9acc225762c4dd4fd04adfe
뒤에 동일한 커밋이 있어서 이 커밋은 신경안쓰셔도 됩니다.

HongJungKim-dev reviewed (commented):
14376da09f4385666b28ca5216d3098e3a4a888b
시간 관계상 여기까지 해야할것 같습니다. 일어나서 다시 하겠습니다.

아직 강의를 완벽하게 실습하지 못했습니다.
그런데 이번주 내로 리뷰도 받으면서 이 과제를 수행하기 위해서 일단 다른분들 PR의 순서를 참고하면서 필요한 부분을 강의를 참고하면서 했습니다. 코드를 작성할때는 물론 스스로 인출중심으로 했습니다.

HongJungKim-dev committed:
Fix error deleteTask case of reducer test

deleteTask타입인 경우 테스트코드에서 toBe가 ''이어야해서 수정합니다.

HongJungKim-dev committed:
Add reducer of adding and deleteing task

addTask와 deleteTask인 경우에 대해 reducer을 구현했습니다.

HongJungKim-dev committed:
Add reducer test code of adding and deleteing task

addTask와 deleteTask인 경우에 대해 reducer을테스트하는 코드를 추가합니다.

HongJungKim-dev committed:
Fix wrong reducer test code of addTask type with taskTitle

addTask에서 taskTitle이 입력되면값이 있고 이후에 공백이 되는 것을
놓쳐서 reducer test code를 바로잡았습니다.

HongJungKim-dev committed:
Add reducer test code in case nothing taskTitle

taskTitle이 없는 경우를 test하는 reducer test code를 추가합니다.

HongJungKim-dev committed:
Refactor action of updating title

taskTitle을 업데이트하는 경우의 액션에 해당하는 함수를 import해서
사용하도록 리팩토링했습니다.

HongJungKim-dev committed:
Fix wrong reducer test code

addTask와 dleteTaask인 경우 테스트 코드가 잘못되어서 수정했습니다.

HongJungKim-dev committed:
Add reducer if taskTitle is nothing

reducer에서 taskTitle이 ''일때 예외처리를 추가했습니다.

HongJungKim-dev committed:
Add addTask function at reducer test code

reducer test code에 addTask function을 추가합니다.

HongJungKim-dev committed:
Add addTask function at actions

addTask함수를 추가하여 actions를 구현합니다.

HongJungKim-dev committed:
Add deleteTask at reducer test code

deleteTask함수를 reducer 테스트 코드로 추가합니다.

HongJungKim-dev committed:
Add deleteTask at action

deleteTask함수를 action에 구현했습니다.

moonkii commented on package.json:
```diff
@@ -41,6 +41,8 @@
+    "redux": "^4.0.5"
   }
-}
+}
```
마지막 빈 줄이 왜 필요한지 제가 설명드린 적이 있나요?

항상 Pull Request를 보내고 나서 자신이 작성한 코드를 Files changed 에서 다시 한번 살펴보는 습관을 들이시길 추천해요.

파일 끝에 개행을 추가하지 않으면 파일 끝에 ⛔️ (No newline at end of file) 경고 문구가 붙어요.

파일 끝에 개행을 추가 하는 이유는 예전에는 컴파일러가 파일 끝에 개행문자가 없으면 한 줄이 끝나지 않은 것으로 인식해서 에러가 발생하는 이슈가 있었기 때문이에요.

최근에는 파일 끝에 개행문자가 없어도 컴파일러에서 별다른 문제가 발생하지 않지만 혹시나 모를 잠재적인 에러나 [POSIX](https://ko.wikipedia.org/wiki/POSIX) 에 명세되어 있기 때문에 개인적으로는 파일 끝에 개행문자를 추가 하시길 권장해요.

HongJungKim-dev replied:
앞으로 FIles changed를 다시 한번 제대로 확인하겠습니다. 혹시 모를 EOF 미인식으로 인한 오류때문에 newline을 추가한다는것 기억하겠습니다!

HongJungKim-dev committed:
Add newLine at end of file

예기치 못한 EOF미인식으로 인한 에러를 피하기 위해 문장의 끝에 새로운 줄을 추가했습니다.

HongJungKim-dev reviewed (commented):
reduer.test.js에서 완료버튼을 선택하기 때문에 없는 아이디를 삭제하는 일은 없을것이라고 판단해서 그 부분은 테스트를 진행하지 않았는데 강의에서는 테스트를 진행했습니다. 저는 아예 필요하다고 생각 못했는데 단지 테스트 하는 사람에 따라 판단이 다른 것일까요?

HongJungKim-dev committed:
Refactor reducer test code because of redundant codes

reducer test code에서 중복된 코드를 reduceAddTask의 함수로 리팩토링합니다.

HongJungKim-dev committed:
Add inintialState at reducer

reducer에서 초기상태를 추가했습니다.

HongJungKim-dev committed:
Add store to manage state

상태를 관리하기 위해서 store를구현했습니다.

HongJungKim-dev committed:
Add provider to access store

App이 redux store를 접근할 수 있도록 Provider를 추가합니다.

HongJungKim-dev committed:
Add InputContainer test code

InputContainer 테스트 코드를 추가합니다.

HongJungKim-dev committed:
Make InputContainer

InputContainer를 기본세팅으로 구현했습니다.

HongJungKim-dev committed:
Add InputContainer's dispatch function test code

InputContainer test code에서 dispatch함수 확인을  추가했습니다.

HongJungKim-dev committed:
Make InputContainer using useDispatch and useSelector

useDispatch와 useSelector을 이용해서 InputContainer를 구현했습니다.

HongJungKim-dev commented:
InputContainer.test.jsx에서 계속 run상태인 이유를 모르겠습니다.
![image](https://user-images.githubusercontent.com/58525009/108506688-8db97a00-72fc-11eb-81ad-17c4e450faaf.png)

HongJungKim-dev committed:
Add ListContainer test code to test if text is null or not

기본으로 null이 아닌지 ListContainer테스트하는 코드를 추가합니다.

HongJungKim-dev committed:
Refactor InputContainer test code with it pattern

it패턴으로 InputContainer 테스트 코드를 리팩토링합니다.

HongJungKim-dev committed:
Add InputContainer test code to test printing elements and calling func

element들이 출력되는지, 입력값 변경시 onChange함수가 호출되는지 확인하는
InputContainer 테스트를 추가합니다.

HongJungKim-dev commented:
App테스트 -> App구현 -> 리팩토링 ->reducer테스트 -> reducer구현 ->리팩토링 -> App의 상태관리를 redux로 대체-> ListContainer와 InputContainer 테스트 ->

HongJungKim-dev committed:
ListContainer을 구현합니다.

HongJungKim-dev committed:
ListContainer을 테스트하는데, 할 일없는 경우를 추가합니다.

HongJungKim-dev committed:
ListContainer를 테스트하는데, 완료버튼을 누른 경우를 추가합니다.

HongJungKim-dev committed:
App의 상태관리를 redux로 대체합니다.

HongJungKim-dev committed:
App을 구현합니다.

HongJungKim-dev committed:
Page와 Page test를 삭제합니다.

HongJungKim-dev commented:
## 미해결 오류

오타났는지 살펴봐도 없는것 같고 왜 오류가 나는지 모르겠습니다. 어떤식으로 접근하는게 좋을지 모르겠습니다.
![image](https://user-images.githubusercontent.com/58525009/108549099-0ede3480-7330-11eb-8b29-5c7dd838733f.png)

HongJungKim-dev commented:
과제1은 일단 여기까지만 마무리하고 리뷰를 기다리겠습니다! 과제2를 조금이라도 건드려야해서 이제 과제2를 시작은 해야할것 같습니다.

HongJungKim-dev committed:
Fix error of call own tag

ListContainer와 InputContainer에서 태그를 스스로 불러서 나는 오류를 각각
List 태그, Input태그로 수정하여 고쳤습니다.

HongJungKim-dev committed:
InputContainer를 리팩토링하여 jest.fn이 여러번 사용되어 생길 오류 방지

HongJungKim-dev committed:
ListContainer를 리팩토링하여 여러 완료버튼에 대해 처리합니다.

HongJungKim-dev commented:
## 미해결 오류

![image](https://user-images.githubusercontent.com/58525009/108559814-2cff6100-733f-11eb-8cfe-833d86806ef6.png)

moonkii commented:
> reduer.test.js에서 완료버튼을 선택하기 때문에 없는 아이디를 삭제하는 일은 없을것이라고 판단해서 그 부분은 테스트를 진행하지 않았는데 강의에서는 테스트를 진행했습니다. 저는 아예 필요하다고 생각 못했는데 단지 테스트 하는 사람에 따라 판단이 다른 것일까요?

사람 마다 테스트를 해야하는 부분에 대한 판단은 다를 수 있어요. 
그래서 테스트 커버리지 100% 가 꼭 좋은 퀄리티를 보장하는게 아니라는 거에요.

moonkii commented:
> App -> reducer -> container 순으로 하는게 아니라 reducer가 맨 마지막인가요?

이미 구현이 되어 있는 상황이었기 때문에 그렇게 설명을 드린거에요. 꼭 정해져 있는건 아니에요.
중요한건 reducer 와 container는 독립적이다 라는 것이죠.
컴포넌트를 먼저 분리하고 나서 Redux 를 추가할 수 있고 Redux 가 추가된 상태에서도 컴포넌트가 분리될 수 있어요.

> 
> ## 올바른 개발 흐름 질문
> 개괄 : App -> reducer -> container
> 
> 구체적 순서 :
> App테스트 -> App구현 -> 리팩토링
> ->reducer테스트 -> reducer구현 ->리팩토링
> -> App의 상태관리를 redux로 대체
> -> ListContainer 테스트 -> ListContainer 구현 ->리팩토링
> -> InputContainer 테스트 -> InputContainer 구현 -> 리팩토링
> 
> 이 과정으로 했어야 하는것 같습니다.
> App테스트를 먼저 해서 이후에 상태관리를 redux로 대체하는것이 더 간단하면서도 redux로 상태관리를 변경한다는 의미에서 자연스러운 흐름인것 같습니다.
> 이것이 맞는지 궁금합니다.

네 말씀하신대로 진행하셔도 좋을 것 같아요.

> ## 구현순서 질문
> ListContainer와 InputContainer는 서로 구현 순서가 상관없다고 생각하는데 맞는것인지 궁금합니다.

네 서로 의존성이 없어서 순서는 상관없어요.

moonkii commented on src/InputContainer.jsx:
```diff
@@ -0,0 +1,33 @@
+    taskTitle: state.title,
+  }));
+
+  function handleChangeTitle(event) {
```
`event` 핸들러를 넘겨주는 이유가 있을까요?

moonkii commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,55 @@
+  const dispatch = jest.fn();
+  beforeEach(() => jest.clearAllMocks());
+
+  useDispatch.mockImplementation(() => dispatch);
+  useSelector.mockImplementation((selector) => selector({
+    taskTitle: '할 일1',
+  }));
```
이 부분은 매번 컴포넌트가 render 될 때 마다 새롭게 필요한 부분이에요.

그래서 `beforeEach` 에 넣어줘야 돼요.

HongJungKim-dev replied:
네 수정하겠습니다.

moonkii commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,55 @@
+      <InputContainer />
+    ));
+
+    expect(getByText(/추가/)).not.toBeNull();
+    expect(getByLabelText('할 일')).not.toBeNull();
+    expect(getByPlaceholderText('할 일을 입력해 주세요')).not.toBeNull();
```
`getBy~` 와 `queryBy~` 에 대해서 찾아보시면 좋을 것 같아요.

HongJungKim-dev replied:
찾는 요소가 없으면 getBy는 오류를 발생시키고 queryBy는 null을 반환시킵니다.
지금처럼 not.toBeNull을 사용하는 경우에는 queryBy가 적절한것 같아서 수정하겠습니다.

참고자료
* https://testing-library.com/docs/dom-testing-library/cheatsheet/

moonkii commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,55 @@
+  });
+
+  it('일정을 추가하는 함수가 실행합니다.', () => {
+    const { getByText } = render((
+      <InputContainer />
+    ));
```
이렇게 간단한 경우에는 한줄로 해줘도 괜찮을 것 같아요.

HongJungKim-dev replied:
네 수정하겠습니다.

moonkii commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,55 @@
+      <InputContainer />
+    ));
+
+    expect(dispatch).not.toBeCalled();
+    fireEvent.click(getByText('추가'));
+    expect(dispatch).toBeCalled();
```
```suggestion
    expect(dispatch).not.toBeCalled();
    
    fireEvent.click(getByText('추가'));
    
    expect(dispatch).toBeCalled();
```

의미상으로 구분되는 경우 빈 줄을 넣어서 구분해주면 좋아요.

HongJungKim-dev replied:
click이벤트와 그에따른 함수호출이라서 어차피 같이 있어야한다고 생각해서 빈줄을 추가하지 않았는데 어떤 의미 때문에 구분해야하는것인지 궁금합니다.

moonkii commented on src/ListContainer.test.jsx:
```diff
@@ -0,0 +1,65 @@
+describe('ListContainer', () => {
+  context('할 일이 없는 경우', () => {
+    it('할 일이 없음을 출력합니다.', () => {
+      useSelector.mockImplementation((selector) => selector({
+        tasks: [],
+      }));
```
매번 반복되는 부분을 어떻게 하면 피할 수 있을까요?

그렇게 되면 `tasks` 가 달라야 하는 상황이 발생하는데 `given2` 라이브러리를 찾아보시면 도움될거에요 ;)

HongJungKim-dev replied:
given2를 찾아서 시도해봤는데 맞는지 모르겠습니다.

참고자료
* https://www.npmjs.com/package/given2/v/1.0.3

moonkii commented:
> => [7576ab6](https://github.com/CodeSoom/react-week4-assignment-1/commit/7576ab6177f375f81e06e5e50d6b89c7e84fa47f) 자기 자신의 태그를 호출해서 run상태라고 파악했었는데 여기서 run이 끝나서 오류를 출력했습니다. 저 오류도 결국 컴포넌트 안에서 자신의 컴포넌트를 정의하지 않은 문제인데 그것이 옳지 않은것이고 Input으로 바꿔야하지만 밑의 경우처럼 run상태에 머무르지 않고 어떻게 오류메시지를 출력할 수 있는지 궁금합니다.

저라면 `run` 이 비정상적으로 안끝나면 어디서 안끝나는지 확인 후 테스트를 중지 시켜서 그 부분을 확인해 볼 것 같아요.

HongJungKim-dev committed:
InputContiner테스트 리팩토링하여 render마다 useDispatch,useSelector새로

HongJungKim-dev committed:
InputContainer테스트 리팩토링하여 getBy를 queryBy로 변경합니다.

HongJungKim-dev committed:
ListContainer를 테스트하는데 given2를 시도했습니다.

moonkii commented on src/InputContainer.test.jsx:
```diff
@@ -21,15 +21,15 @@ describe('InputContainer', () => {
-    expect(getByText(/추가/)).not.toBeNull();
+    expect(queryByText(/추가/)).not.toBeNull();
     expect(getByLabelText('할 일')).not.toBeNull();
     expect(getByPlaceholderText('할 일을 입력해 주세요')).not.toBeNull();
```
이 부분은 따로 `getBy~` 를 사용하는 이유가 있나요?

HongJungKim-dev replied:
이부분을 수정 못했던것 같습니다. queryBy가 적절한것은 모두 찾아서 수정하겠습니다.

moonkii commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,56 @@
+  });
+
+  it('입력된 값과 element들을 출력합니다.', () => {
+    const { queryByText, getByLabelText, getByPlaceholderText } = render((<InputContainer />));
```
```suggestion
    const { queryByText, getByLabelText, getByPlaceholderText } = render(<InputContainer />);
```

한 줄일 때는 괄호가 하나면 충분해요 ;)

HongJungKim-dev replied:
여러 줄일때는 괄호를 두번 하는 이유가 무엇인지 모르겠습니다.
stackoverflow에 render double parentheses라고 검색했는데 적절한 답변을 못찾겠습니다. 어떤 부분을 다시 봐야할까요?

HongJungKim-dev committed:
리팩토링하여 queryBy가 적절한 모든 코드를 getBy에서 queryBy로 수정했습니다.
